File: ts/tests/nepaliCalendar.test.ts
import { describe, it, expect, beforeAll } from 'vitest';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import {
  initializeCalendar,
  convertADToBS,
  convertBSToAD,
  isValidBSDate,
  getDaysInBSMonth,
  getTotalDaysInBSYear,
  toNepaliDigits,
  formatNepaliDate,
  formatEnglishDate,
  getNepaliMonthName,
  getRelativeTimeInNepali,
  toNepalTime,
  fromNepalTime,
} from '../src/index.js';
import { InlineCalendarDataSource } from '../src/index.js';
import type { CalendarJson } from '../src/index.js';

// Load full calendar JSON (BS 2000–2100) so conversion tests work correctly.
// The conversion engine iterates from BS 2000, so all intermediate years are required.
const calendarJson: CalendarJson = JSON.parse(
  readFileSync(join(dirname(fileURLToPath(import.meta.url)), '../../data/nepali_calendar_data.json'), 'utf8')
);

beforeAll(() => {
  const source = new InlineCalendarDataSource(
    calendarJson.years,
    calendarJson.supportedRange.start,
    calendarJson.supportedRange.end,
  );
  initializeCalendar(source);
});

describe('convertADToBS', () => {
  it('converts AD 2024-04-12 to BS 2080/12/30', () => {
    const result = convertADToBS(new Date(2024, 3, 12));
    expect(result).toEqual({ year: 2080, month: 12, day: 30 });
  });

  it('converts AD 2024-04-13 to BS 2081/1/1 (new year)', () => {
    const result = convertADToBS(new Date(2024, 3, 13));
    expect(result).toEqual({ year: 2081, month: 1, day: 1 });
  });

  it('converts AD 2025-09-17 to BS 2082/6/1', () => {
    const result = convertADToBS(new Date(2025, 8, 17));
    expect(result).toEqual({ year: 2082, month: 6, day: 1 });
  });
});

describe('convertBSToAD', () => {
  it('converts BS 2081/1/1 to AD 2024-04-13', () => {
    const result = convertBSToAD(2081, 1, 1);
    expect(result.getFullYear()).toBe(2024);
    expect(result.getMonth()).toBe(3);
    expect(result.getDate()).toBe(13);
  });

  it('converts BS 2080/12/30 to AD 2024-04-12', () => {
    const result = convertBSToAD(2080, 12, 30);
    expect(result.getFullYear()).toBe(2024);
    expect(result.getMonth()).toBe(3);
    expect(result.getDate()).toBe(12);
  });

  it('round-trips: AD→BS→AD', () => {
    const original = new Date(2025, 0, 15);
    const bs = convertADToBS(original);
    const back = convertBSToAD(bs.year, bs.month, bs.day);
    expect(back.getFullYear()).toBe(original.getFullYear());
    expect(back.getMonth()).toBe(original.getMonth());
    expect(back.getDate()).toBe(original.getDate());
  });

  it('throws on invalid BS date', () => {
    expect(() => convertBSToAD(2082, 13, 1)).toThrow();
    expect(() => convertBSToAD(2082, 1, 32)).toThrow();
  });
});

describe('isValidBSDate', () => {
  it('accepts valid dates', () => {
    expect(isValidBSDate(2082, 1, 1)).toBe(true);
    expect(isValidBSDate(2082, 1, 31)).toBe(true);
  });

  it('rejects out-of-range month', () => {
    expect(isValidBSDate(2082, 0, 1)).toBe(false);
    expect(isValidBSDate(2082, 13, 1)).toBe(false);
  });

  it('rejects day exceeding month length', () => {
    expect(isValidBSDate(2082, 1, 32)).toBe(false);
  });

  it('rejects unknown year', () => {
    expect(isValidBSDate(1999, 1, 1)).toBe(false);
  });
});

describe('getDaysInBSMonth', () => {
  it('returns correct days for BS 2082/1 (Baisakh)', () => {
    expect(getDaysInBSMonth(2082, 1)).toBe(31);
  });

  it('returns correct days for BS 2082/11 (Falgun)', () => {
    expect(getDaysInBSMonth(2082, 11)).toBe(30);
  });
});

describe('getTotalDaysInBSYear', () => {
  it('returns 365 or 366 for each known year', () => {
    const total = getTotalDaysInBSYear(2082);
    expect(total).toBeGreaterThanOrEqual(365);
    expect(total).toBeLessThanOrEqual(366);
  });
});

describe('toNepaliDigits', () => {
  it('converts 0 to ०', () => expect(toNepaliDigits(0)).toBe('०'));
  it('converts 2082 to २०८२', () => expect(toNepaliDigits(2082)).toBe('२०८२'));
  it('converts 15 to १५', () => expect(toNepaliDigits(15)).toBe('१५'));
});

describe('formatNepaliDate', () => {
  it('formats in English', () => {
    expect(formatNepaliDate({ year: 2082, month: 11, day: 8 }, 'en')).toBe('Falgun 8, 2082');
  });

  it('formats in Nepali', () => {
    expect(formatNepaliDate({ year: 2082, month: 11, day: 8 }, 'ne')).toBe('फागुन 8, 2082');
  });
});

describe('formatEnglishDate', () => {
  it('formats correctly', () => {
    expect(formatEnglishDate(new Date(2024, 3, 14))).toBe('April 14, 2024');
  });
});

describe('getNepaliMonthName', () => {
  it('returns Baisakh for month 1 in English', () => {
    expect(getNepaliMonthName(1, 'en')).toBe('Baisakh');
  });

  it('returns बैशाख for month 1 in Nepali', () => {
    expect(getNepaliMonthName(1, 'ne')).toBe('बैशाख');
  });
});

describe('toNepalTime / fromNepalTime', () => {
  it('shifts by +5:45', () => {
    const utc = new Date('2025-01-01T00:00:00Z');
    const nep = toNepalTime(utc);
    expect(nep.getUTCHours()).toBe(5);
    expect(nep.getUTCMinutes()).toBe(45);
  });

  it('round-trips', () => {
    const utc = new Date('2025-06-15T10:30:00Z');
    expect(fromNepalTime(toNepalTime(utc)).getTime()).toBe(utc.getTime());
  });
});

describe('getRelativeTimeInNepali', () => {
  it('returns हालै for null', () => {
    expect(getRelativeTimeInNepali(null)).toBe('हालै');
  });

  it('returns seconds ago for recent date', () => {
    const recent = new Date(Date.now() - 30_000);
    const result = getRelativeTimeInNepali(recent);
    expect(result).toContain('सेकेण्ड अगाडि');
  });

  it('returns minutes ago', () => {
    const past = new Date(Date.now() - 10 * 60_000);
    expect(getRelativeTimeInNepali(past)).toContain('मिनेट अगाडि');
  });
});

File: dart/test/nepali_calendar_test.dart
import 'package:jhigu_patro_utils/jhigu_patro_utils.dart';
import 'package:test/test.dart';

const Map<int, List<int>> _years = {
  2079: [31, 32, 31, 32, 31, 30, 30, 29, 30, 29, 30, 30],
  2080: [31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 30],
  2081: [31, 32, 31, 32, 31, 30, 30, 30, 29, 30, 29, 31],
  2082: [31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
};

void main() {
  setUpAll(() {
    NepaliCalendar.initialize(
      InlineCalendarDataSource(years: _years, supportedStartYear: 2079, supportedEndYear: 2082),
    );
  });

  group('convertADToBS', () {
    test('AD 2024-04-12 → BS 2080/12/30', () {
      expect(
        NepaliCalendar.convertADToBS(DateTime.utc(2024, 4, 12)),
        equals(NepaliDate(year: 2080, month: 12, day: 30)),
      );
    });

    test('AD 2024-04-13 → BS 2081/1/1 (new year)', () {
      expect(
        NepaliCalendar.convertADToBS(DateTime.utc(2024, 4, 13)),
        equals(NepaliDate(year: 2081, month: 1, day: 1)),
      );
    });

    test('AD 2025-09-17 → BS 2082/6/1', () {
      expect(
        NepaliCalendar.convertADToBS(DateTime.utc(2025, 9, 17)),
        equals(NepaliDate(year: 2082, month: 6, day: 1)),
      );
    });
  });

  group('convertBSToAD', () {
    test('BS 2081/1/1 → AD 2024-04-13', () {
      final result = NepaliCalendar.convertBSToAD(2081, 1, 1);
      expect(result.year, 2024);
      expect(result.month, 4);
      expect(result.day, 12);
    });

    test('BS 2080/12/30 → AD 2024-04-12', () {
      final result = NepaliCalendar.convertBSToAD(2080, 12, 30);
      expect(result.year, 2024);
      expect(result.month, 4);
      expect(result.day, 12);
    });

    test('round-trip AD→BS→AD', () {
      final original = DateTime.utc(2025, 1, 15);
      final bs = NepaliCalendar.convertADToBS(original);
      final back = NepaliCalendar.convertBSToAD(bs.year, bs.month, bs.day);
      expect(back.year, original.year);
      expect(back.month, original.month);
      expect(back.day, original.day);
    });

    test('throws on invalid date', () {
      expect(() => NepaliCalendar.convertBSToAD(2082, 13, 1), throwsArgumentError);
      expect(() => NepaliCalendar.convertBSToAD(2082, 1, 32), throwsArgumentError);
    });
  });

  group('isValidBSDate', () {
    test('accepts valid date', () => expect(NepaliCalendar.isValidBSDate(2082, 1, 1), isTrue));
    test('rejects month 0',    () => expect(NepaliCalendar.isValidBSDate(2082, 0, 1), isFalse));
    test('rejects month 13',   () => expect(NepaliCalendar.isValidBSDate(2082, 13, 1), isFalse));
    test('rejects day > month length', () => expect(NepaliCalendar.isValidBSDate(2082, 1, 32), isFalse));
    test('rejects unknown year',       () => expect(NepaliCalendar.isValidBSDate(1999, 1, 1), isFalse));
  });

  group('getDaysInBSMonth', () {
    test('2082 Baisakh = 31 days', () => expect(NepaliCalendar.getDaysInBSMonth(2082, 1), 31));
    test('2082 Falgun = 30 days',  () => expect(NepaliCalendar.getDaysInBSMonth(2082, 11), 30));
  });

  group('toNepaliDigits', () {
    test('0 → ०',       () => expect(NepaliCalendar.toNepaliDigits(0), '०'));
    test('2082 → २०८२', () => expect(NepaliCalendar.toNepaliDigits(2082), '२०८२'));
    test('15 → १५',     () => expect(NepaliCalendar.toNepaliDigits(15), '१५'));
  });

  group('formatNepaliDate', () {
    test('English format', () {
      expect(
        NepaliCalendar.formatNepaliDate(NepaliDate(year: 2082, month: 11, day: 8)),
        'Falgun 8, 2082',
      );
    });
    test('Nepali format', () {
      expect(
        NepaliCalendar.formatNepaliDate(NepaliDate(year: 2082, month: 11, day: 8), nepali: true),
        'फागुन 8, 2082',
      );
    });
  });

  group('toNepalTime / fromNepalTime', () {
    test('shifts UTC by +5:45', () {
      final utc = DateTime.utc(2025, 1, 1, 0, 0);
      final nep = NepaliCalendar.toNepalTime(utc);
      expect(nep.hour, 5);
      expect(nep.minute, 45);
    });

    test('round-trips', () {
      final utc = DateTime.utc(2025, 6, 15, 10, 30);
      expect(NepaliCalendar.fromNepalTime(NepaliCalendar.toNepalTime(utc)), equals(utc));
    });
  });

  group('getRelativeTimeInNepali', () {
    test('null → हालै', () => expect(NepaliCalendar.getRelativeTimeInNepali(null), 'हालै'));

    test('30 seconds ago → सेकेण्ड अगाडि', () {
      final recent = DateTime.now().subtract(const Duration(seconds: 30));
      expect(NepaliCalendar.getRelativeTimeInNepali(recent), contains('सेकेण्ड अगाडि'));
    });

    test('10 minutes ago → मिनेट अगाडि', () {
      final past = DateTime.now().subtract(const Duration(minutes: 10));
      expect(NepaliCalendar.getRelativeTimeInNepali(past), contains('मिनेट अगाडि'));
    });
  });
}
